shared-module/tilepalettemapper/TilePaletteMapper.c

```diff
@@ -0,0 +1,78 @@
+// This file is part of the CircuitPython project: https://circuitpython.org
+//
+// SPDX-FileCopyrightText: Copyright (c) 2025 Tim Cocks for Adafruit Industries
+//
+// SPDX-License-Identifier: MIT
+#include <stdlib.h>
+#include "py/runtime.h"
+#include "shared-bindings/tilepalettemapper/TilePaletteMapper.h"
+#include "shared-bindings/displayio/Palette.h"
+
+void common_hal_tilepalettemapper_tilepalettemapper_construct(tilepalettemapper_tilepalettemapper_t *self,
+    mp_obj_t palette, uint16_t input_color_count, uint16_t width, uint16_t height) {
+
+    self->palette = palette;
+    self->width_in_tiles = width;
+    self->height_in_tiles = height;
+    self->input_color_count = input_color_count;
+    self->needs_refresh = false;
+    int mappings_len = width * height;
+    self->tile_mappings = (uint16_t **)m_malloc(mappings_len * sizeof(uint16_t *));
+    for (int i = 0; i < mappings_len; i++) {
+        self->tile_mappings[i] = (uint16_t *)m_malloc(input_color_count * sizeof(uint16_t));
+        for (uint16_t j = 0; j < input_color_count; j++) {
+            self->tile_mappings[i][j] = j;
+        }
+    }
+}
+
+uint16_t common_hal_tilepalettemapper_tilepalettemapper_get_width(tilepalettemapper_tilepalettemapper_t *self) {
+    return self->width_in_tiles;
+}
+
+uint16_t common_hal_tilepalettemapper_tilepalettemapper_get_height(tilepalettemapper_tilepalettemapper_t *self) {
+    return self->height_in_tiles;
+}
+
+mp_obj_t common_hal_tilepalettemapper_tilepalettemapper_get_palette(tilepalettemapper_tilepalettemapper_t *self) {
+    return self->palette;
+}
+
+mp_obj_t common_hal_tilepalettemapper_tilepalettemapper_get_mapping(tilepalettemapper_tilepalettemapper_t *self, uint16_t x, uint16_t y) {
+    int index = x + y * self->width_in_tiles;
+    mp_obj_t result[self->input_color_count];
+    for (uint32_t i = 0; i < self->input_color_count; i++) {
+        result[i] = mp_obj_new_int(self->tile_mappings[index][i]);
+    }
+    return mp_obj_new_tuple(self->input_color_count, result);
+}
+
+void common_hal_tilepalettemapper_tilepalettemapper_set_mapping(tilepalettemapper_tilepalettemapper_t *self, uint16_t x, uint16_t y, size_t len, mp_obj_t *items) {
+    uint32_t palette_len = common_hal_displayio_palette_get_len(self->palette);
+    for (uint16_t i = 0; i < len; i++) {
```

Comment:
Make sure you can't assign more than input color count.

```suggestion
    for (uint16_t i = 0; i < MIN(len, self->input_color_count); i++) {
```